### PR TITLE
tackle-75-business-logic-for-create-assessment

### PR DESCRIPTION
**Pull request by jonathanvila**
Issue : https://github.com/konveyor/tackle-pathfinder/issues/11

### Features included : 
* POST assessments operation now will copy the default (first one) Questionnaire structure ( categories, questions, options ) into the assessment
* Old Pathfinder questionnaire inserted  as the default questionnaire

### Pre steps : 
1. Deploy tackle-controls ( to have the keycloak instance running ) on your K8s cluster
```Shell
$ kubectl apply -f { tackle-controls kubernetes deployment file }
```
2. Edit the Ingress object to allow direct request to Pathfinder API
```Shell
$ kubectl edit ingress tackle -n tackle
```
Add this below `paths:`
```Yaml
     - backend:
          service:
            name: tackle-pathfinder
            port:
              number: 8080 
        path: /pathfinder
        pathType: ImplementationSpecific
```
3. Build and push 
```Shell
$ ./mvnw -U -B package -Dquarkus.container-image.push=true -Dquarkus.container-image.group={your quay user} -Dquarkus.container-image.registry=quay.io -Dquarkus.container-image.username={your quay user} -Dquarkus.container-image.password={your quay pwd} -Pnative
```
4. Change the image in the deployment file (src/main/kubernetes/tackle-pathfinder.yaml) line 224 to be the one you have pushed
5. Deploy tackle-pathfinder on your K8s cluster , same namespace as tackle-controls
```Shell
$ kubectl apply -f src/main/kubernetes/tackle-pathfinder.yaml 
```

### Test case 1 : Obtain Keycloak secret
```Shell
$ export access_token=$(\
    curl -X POST http://{cluster IP}/auth/realms/quarkus/protocol/openid-connect/token \
    --user backend-service:secret \
    -H 'content-type: application/x-www-form-urlencoded' \
    -d 'username=alice&password=alice&grant_type=password' | jq --raw-output '.access_token' \
 )
```
> Expectation test case 1
```Shell
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1825  100  1776  100    49  18500    510 --:--:-- --:--:-- --:--:-- 19010
```

### Test case 2 : Create assessment for applicationId=20
```Shell
curl 'http://{cluster IP}/pathfinder/assessments' \
  -H 'Content-Type: application/json' -H 'Accept: application/json' -H "Authorization: Bearer "$access_token \
  -d "{ \"applicationId\": 20 }" \
  -s -i
```
> Expectation Test case 2
```Shell
HTTP/1.1 201 Created
Date: Mon, 22 Mar 2021 11:05:29 GMT
Content-Type: application/json
Content-Length: 46
Connection: keep-alive

{"id":3,"applicationId":28,"status":"STARTED"}

```
### Test case 3 : DB check for assessment for applicationId=20
1. create a postgre client pod with 
```Shell
$ kubectl run pgcli --image riuvshin/postgres-client
```
2. check status of the pod, and leave ( ^c ) when Ready
```Shell
$ kubectl get pod pgcli -w 
```
3. sh into the pod  
```Shell
$ kubectl exec -ti pgcli sh
```
4. connect to the DB
```Shell
$ psql --host=pathfinder-postgresql --port=5432 --username=pathfinder --dbname=pathfinder_db
```
5. execute query
```Shell
$ select count(*) from assessment_singleoption;
```
 
> Expectation Test case 3
```Shell
pathfinder_db=# select count(*) from single_option ;
 count 
-------
   174
(1 row)
```

### Test case 4 : Simultaneous calls to create assessment return 201 for the first and 400 for the rest ( to avoid multiple assessments per application )
1. create a file with this content
```Shell
export access_token=$(\
    curl -X POST http://$2/auth/realms/quarkus/protocol/openid-connect/token \
    --user backend-service:secret \
    -H 'content-type: application/x-www-form-urlencoded' \
    -d 'username=alice&password=alice&grant_type=password' | jq --raw-output '.access_token' \
 )
time
curl 'http://$2/pathfinder/assessments' \
  -H 'Content-Type: application/json' -H 'Accept: application/json' -H "Authorization: Bearer "$access_token \
  -d "{ \"applicationId\": $1 }" \
  -v -s &
time
curl 'http://$2/pathfinder/assessments' \
  -H 'Content-Type: application/json' -H 'Accept: application/json' -H "Authorization: Bearer "$access_token \
  -d "{ \"applicationId\": $1 }" \
  -v -s &
time
curl 'http://$2/pathfinder/assessments' \
  -H 'Content-Type: application/json' -H 'Accept: application/json' -H "Authorization: Bearer "$access_token \
  -d "{ \"applicationId\": $1 }" \
  -v -s &
```
2. execute it with `bash -x file.sh {incremental number}  {cluster ip} `
If you are using minikube you can get your cluster ip with : minikube ip

> Expectation Test case 4
Something similar to this ( you have 1 HTTP/1.1 201 , and 2 HTTP/1.1 400 )
```Shell
< HTTP/1.1 201 Created
< Date: Tue, 13 Apr 2021 08:17:53 GMT
< Content-Type: application/json
< Content-Length: 50
< Connection: keep-alive
< 
* Connection #0 to host 192.168.49.2 left intact
{"id":1051,"applicationId":802,"status":"STARTED"}* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< Date: Tue, 13 Apr 2021 08:17:53 GMT
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host 192.168.49.2 left intact
* Mark bundle as not supporting multiuse
< HTTP/1.1 400 Bad Request
< Date: Tue, 13 Apr 2021 08:17:53 GMT
< Content-Length: 0
< Connection: keep-alive
< 
* Connection #0 to host 192.168.49.2 left intact

```

### CODE REVIEW

File: src/main/java/io/tackle/pathfinder/model/AbstractEntity.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        if (Objects.nonNull(context)) {
+            String username = context.getPrincipal().getName();
+            if (Objects.nonNull(username)) {
+                return username;
+            }
+        }
+        // since all service are authenticated, it should never get here
+        // so maybe worth evaluating to throw an exception? food for thoughts
+        return "";
+    }
```
Comment (carlosthe19916) [inline src/main/java/io/tackle/pathfinder/model/AbstractEntity.java:61]: - From the comment you mentioned in the commented lines I've also read https://github.com/quarkusio/quarkus/issues/6948#issuecomment-766184194 and https://github.com/quarkusio/quarkus/issues/6948#issuecomment-766367218 and It makes me wonder how do we know this block of code is working? I didn't see any of the tests are checking this value.
- Also I might be wrong but https://github.com/quarkusio/quarkus/issues/6948#issuecomment-766367218 mentions `I think Quarkus is like "you don't use the bean? I'll not pack it then" on buildtime`  so it makes me wonder if this block of code won't work specially in native mode

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 2]
```diff
@@ -12,14 +20,15 @@
...
+import java.util.ArrayList;
 import java.util.List;
 import java.util.Optional;
 
 @ApplicationScoped
 @Log
 public class AssessmentSvc {
     @Inject
-    AssessmentMapper mapper;
+    public AssessmentMapper mapper;
```
Comment (carlosthe19916) [inline src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:31]: Just for my understanding, why did you add  `public`? any special reason?

File: src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    create table assessment_category (
+       id int8 not null,
+        createTime timestamp,
+        createUser varchar(255),
+        deleted boolean not null default false,
+        updateTime timestamp,
+        updateUser varchar(255),
+        name varchar(255),
+        category_order int4,
+        questionnaire_id int8,
```
Comment (carlosthe19916) [inline src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql:24]: This field is confusing since it makes me think that `assessment_category` has a relation with the table `questionnaire` but in fact, the relationship is with the table `assessment_questionnaire`.

```suggestion
        assessment_questionnaire_id int8,
```

File: src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        updateTime timestamp,
+        updateUser varchar(255),
+        language_code varchar(255) not null,
+        name varchar(255) not null,
+        assessment_id int8 not null,
+        questionnaire_id int8 not null,
+        primary key (id)
+    );
+    
+    create table assessment_category (
```
Comment (carlosthe19916) [inline src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql:15]: I can not see any field for saving the user comments. The UI mockups suggest having 1 comment per category.

**Review by carlosthe19916 — CHANGES_REQUESTED** [review]
@jonathanvila I added some comments below and besides my comments I'd like to also mention:
- I've seen the field `notes` in the Assessment entity; `comment` in `AssessmentQuestion` but the `comments/notes` should be located in the `AssessmentCategory`. Could we have a single term for both `comment` and `notes`? unless they both mean different things.
- Before this PR the creation of an assessment was very quick but with this PR the creation of an assessment takes longer (I guess it is because it needs to copy the questionnaire). Since it takes longer for the backend to process the `POST /assessment` request I was able to discover that we can create multiple assessments for a single application which I think is breaking the constraint of having a single assessment per application. See the video below (note that in my video I'm using the UI but you can use curl to have the same results). In the video I'm clicking multiple times the "assess" button.

https://user-images.githubusercontent.com/2582866/114160492-42e4d780-9927-11eb-9ffe-b5ef19b249ad.mp4

Of course, the UI should also disable the button "assess" while executing `POST /assessment` so this is something to enhance also in the UI.

- Could you please format to the .SQL files? I guess it is a matter of using the IDE format feature.

File: src/main/java/io/tackle/pathfinder/model/AbstractEntity.java (hunk 1 above)
Comment (carlosthe19916) [reply src/main/java/io/tackle/pathfinder/model/AbstractEntity.java:61]: I've just tested the native artifact and it seems that `SecurityIdentity context = CDI.current().select(SecurityIdentity.class).get();` works as expected... However, in order to verify this I needed to manually run the native artifact and then go to see the DB manually, In my opinion, throwing an exception makes more sense since we won't notice if something is not working as expected if we just use `return "";`

**Comment by jonathanvila** [comment]
> @jonathanvila I added some comments below and besides my comments I'd like to also mention:
> 
> * I've seen the field `notes` in the Assessment entity; `comment` in `AssessmentQuestion` but the `comments/notes` should be located in the `AssessmentCategory`. Could we have a single term for both `comment` and `notes`? unless they both mean different things.

Well, a comment and notes are not exactly the same and that was my intention. But I see your point, and as it's a matter of perception , I will rename to Comment everywhere as you point.

**Comment by jonathanvila** [comment]
> * Before this PR the creation of an assessment was very quick but with this PR the creation of an assessment takes longer (I guess it is because it needs to copy the questionnaire). Since it takes longer for the backend to process the `POST /assessment` request I was able to discover that we can create multiple assessments for a single application which I think is breaking the constraint of having a single assessment per application. See the video below (note that in my video I'm using the UI but you can use curl to have the same results). In the video I'm clicking multiple times the "assess" button.
>  Peek.2021-04-09.11-29.mp4 
> Of course, the UI should also disable the button "assess" while executing `POST /assessment` so this is something to enhance also in the UI.

:+1:  Good spot .

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 2 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:31]: To be honest, I've been experiencing few times that Quarkus was complaining about the injection of this AssessmentMapper, and apparently it got solved by making it public.

File: src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql (hunk 4 above)
Comment (jonathanvila) [reply src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql:15]: That has been added in the next PR for the GET opperation.

File: src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql (hunk 3 above)
Comment (jonathanvila) [reply src/main/resources/db/migration/V0.0.2_002__create_assessment_quest.sql:24]: OK

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 2 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:31]: But it has to be another thing because leaving it as package now it works.... I experience some random issues with CDI and Quarkus....

File: src/main/java/io/tackle/pathfinder/model/AbstractEntity.java (hunk 1 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/model/AbstractEntity.java:61]: > * From the comment you mentioned in the commented lines I've also read [quarkusio/quarkus#6948 (comment)](https://github.com/quarkusio/quarkus/issues/6948#issuecomment-766184194) and [quarkusio/quarkus#6948 (comment)](https://github.com/quarkusio/quarkus/issues/6948#issuecomment-766367218) and It makes me wonder how do we know this block of code is working? I didn't see any of the tests are checking this value.
This class is a copy from tackle-controls:AbstractEntity , with only the cange on the expression for the delete.

> * Also I might be wrong but [quarkusio/quarkus#6948 (comment)](https://github.com/quarkusio/quarkus/issues/6948#issuecomment-766367218) mentions `I think Quarkus is like "you don't use the bean? I'll not pack it then" on buildtime`  so it makes me wonder if this block of code won't work specially in native mode
I have not paid attention to this class , tbh as it's the same as in TackleControls.
The only change is for the delete expression. We can later discuss about touching or not the tackle-controls one and remove this class from here

Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/model/AbstractEntity.java:61]: In fact, in order to avoid code not tested in this project, I will revert to the tackle-controls-rest  file, and leave the delete expression on the sql files

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java [hunk 5]
```diff
@@ -29,17 +39,65 @@
     @Transactional
     public AssessmentHeaderDto createAssessment(@NotNull Long applicationId) {
         long count = Assessment.count("application_id", applicationId);
-        log.info("Assessments for application_id [ " + applicationId + "] : " + count);
+        log.info("Assessments (" + LocalTime.now() + ") for application_id [ " + applicationId + "] : " + count);
```
Comment (carlosthe19916) [inline src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:42]: Is it necessary to add `LocalTime.now()`? In fact, I'm seeing that many of the `log` messages have the `INFO` level and it means that the production container will continually print every single `log.info()` to the log which might somehow dirty the log output. What do you think? should we change the level of the log from `INFO` to maybe `DEBUG`? if so, should we do it in this PR or in another PR?

**Review by carlosthe19916 — COMMENTED** [review]
@jonathanvila I added just an additional comment to this PR, feel free to take any action about it.

I've tested the latest commit available and the problem of having multiple assessments for an application seems to be solved. I'm sharing the video below of one test I've done using JMeter. This is what I understand from the current status of this PR:

- The problem of multiple assessments per application was solved by adding a `UNIQUE` constraint to the table `assessment`. In this way, we are sure that even if the backend tries to create multiple assessments per application the DB won't allow it to happen.
- The line `assessment.persistAndFlush()` located at https://github.com/jonathanvila/tackle-pathfinder/blob/tackle-75-business-logic-for-create-assessment/src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java#L47 helps to early reject the creation of a new assessment. I've started `pathfinder` using `quarkus:dev` and with `assessment.persistAndFlush()` it takes around 4 seconds for an assessment to be created; on the other hand, using `assessment.persist()` takes around 10 seconds. The timing might change on a native container.


https://user-images.githubusercontent.com/2582866/114537926-15639b00-9c53-11eb-8761-f75bf2a56dd3.mp4

File: src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java (hunk 5 above)
Comment (jonathanvila) [reply src/main/java/io/tackle/pathfinder/services/AssessmentSvc.java:42]: You are right, no need for the LocalTime.now as the log already has it. This came from an old System.out.println
I will remove these logs
:+1:

**Review by carlosthe19916 — APPROVED** [review]

**Comment by carlosthe19916** [comment]
Thanks @jonathanvila , this PR looks good to me.
